### PR TITLE
Allow downloading from Artifactory in npm scripts

### PR DESCRIPTION
**Pull request by yahavi**
- [x] All [tests](https://github.com/jfrog/jfrog-cli-core#tests) passed. If this feature is not already covered by the tests, I added new tests.
- [x] This pull request is on the dev branch.
- [x] I used gofmt for formatting the code before submitting the pull request.
-----

Fix https://github.com/jfrog/jfrog-cli/issues/1347
Fix https://github.com/jfrog/jfrog-cli/issues/1304
Dependency for https://github.com/jfrog/jfrog-cli/pull/1389

In some cases, the npm scripts such as `postinstall` script run `npm install` commands inside a different directory. In such cases, npm sets the `npm_config_registry` environment variable, but not the `npm_config__auth` environment variable. As a result, we may get 401 errors.

### Steps to Reproduce:
**Configure environment**
`package.json`:
```json
{
  "name": "a",
  "version": "v1.0.0",
  "scripts": {
    "postinstall": "npm i --prefix subdir"
  }
}
```
`subdir/package.json`:
```json
{
  "name": "b",
  "version": "v1.0.0",
  "dependencies": {
    "chai": "4.3.4"
  }
}
```
Execute CLI on the root workspace
```
jf npm i
```

### The Fix
Instead of explicitly setting the `_auth` value in the `.npmrc`, do the following:
1. Set `npm_config__auth` environment variable
2. Set `_auth=$JFROG_NPM_AUTH` value in the .npmrc and set JFROG_NPM_AUTH env var. This way we will get a better security as suggested in https://docs.npmjs.com/using-private-packages-in-a-ci-cd-workflow#set-the-token-as-an-environment-variable-on-the-cicd-server.

### Caveat
Since we can't predict on which directories the npm install scripts would run, we can't know on which directories to run `npm ls` and as a result, build info will be collected only on the package.json file in the root workspace.

### CODE REVIEW

**Comment by Hi-Fi** [comment]
`npm_config__auth` works fine at least already with npm6, but might require `nom_config_always_auth` with it, too.

**Comment by yahavi** [comment]
@Hi-Fi,
Thanks for your input. I really appreciate it!

> npm_config__auth works fine

According to npm documentation `npm_config__auth` environment variable exist in npm 7: 
https://docs.npmjs.com/cli/v7/using-npm/config#_auth
But I couldn't find it in npm 6:
https://docs.npmjs.com/cli/v6/using-npm/config

To make sure that it supported on npm 7 and above, I made a full reproducer: https://github.com/yahavi/npm-postinstall-reproducer

Here's the output:
```
➜  npm-postinstall-reproducer git:(main) ✗ nvm use node 12
Now using node v12.22.9 (npm v6.14.15)
➜  npm-postinstall-reproducer git:(main) ✗ ~/code/jfrog-cli/jf npm i
[Info] Running npm install.
npm WARN read-shrinkwrap This version of npm is compatible with lockfileVersion@1, but package-lock.json was generated for lockfileVersion@2. I'll try to do my best with it!

> a@1.0.0 postinstall /Users/yahavi/code/npm-postinstall-reproducer
> npm i --prefix subdir

npm ERR! code E401
npm ERR! Unable to authenticate, need: Basic realm="Artifactory Realm"

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/yahavi/.npm/_logs/2022-01-19T06_20_29_161Z-debug.log
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! a@1.0.0 postinstall: `npm i --prefix subdir`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the a@1.0.0 postinstall script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
npm WARN Local package.json exists, but node_modules missing, did you mean to install?

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/yahavi/.npm/_logs/2022-01-19T06_20_29_240Z-debug.log
[Error] exit status 1
➜  npm-postinstall-reproducer git:(main) ✗ rm -rf package-lock.json subdir/package-lock.json                                   
➜  npm-postinstall-reproducer git:(main) ✗ nvm use node 16                                  
Now using node v16.13.2 (npm v8.1.2)
➜  npm-postinstall-reproducer git:(main) ✗ ~/code/jfrog-cli/jf npm i                        
[Info] Running npm install.

> a@v1.0.0 postinstall
> npm i --prefix subdir


added 7 packages in 4s

up to date, audited 1 package in 5s

found 0 vulnerabilities
[Info] npm install finished successfully.
```


> might require nom_config_always_auth with it, too.

`alway_auth` is populated in the `.npmrc` that we create after running the `jf npm install` command.

Please let me know if I missed something.

**Comment by Hi-Fi** [comment]
NPM@6 documentation is a bit differently built, it just mention all environment variables in one block: https://docs.npmjs.com/cli/v6/using-npm/config#environment-variables

`_auth` is not explicitly mentioned (as there's no config to be called with that, it just generates to `.npmrc` with `login` or `adduser`. But the environment variables part states that all values that go to `.npmrc` are usable also through env variables with just hyphen/underscore modification.

**Comment by yahavi** [comment]
Thanks, @Hi-Fi.
The check was with npm v6.14.15 on MacOS. 
I really hope that I'm wrong, but npm simply ignored the `npm_config__auth` environment variable on my env. With npm 7 and 8, everything worked as expected.

**Comment by Hi-Fi** [comment]
We use that environment variable configuration in AWS without generating even .npmrc against registry in Artifactory. So only NPM configurations in env are:
```
NPM_CONFIG_ALWAYS_AUTH=true
NPM_CONFIG_REGISTRY=<private articatory>
NPM_CONFIG__AUTH=<base64 encoded username:password>
```

AWS uses by default Node 14 and npm 6 in CodeBuild.

**Comment by yahavi** [comment]
@Hi-Fi,
You are right. I changed `npm_config__auth` to `NPM_CONFIG__AUTH` and it worked like a charm on npm 6!
Thanks a lot for this! It helped!

File: artifactory/commands/npm/installorci_test.go [hunk 1]
```diff
@@ -1,15 +1,19 @@
...
 	"strings"
 	"testing"
 
+	biutils "github.com/jfrog/build-info-go/utils"
+	testsutils "github.com/jfrog/jfrog-client-go/utils/tests"
+
 	"github.com/stretchr/testify/assert"
 )
 
+const authToken = "YWRtaW46QVBCN1ZkZFMzN3NCakJiaHRGZThVb0JlZzFl"
```
Comment (Hi-Fi) [inline artifactory/commands/npm/installorci_test.go:15]: This isn't actually token, as authTokens are also supported against specific repositories/scopes.

Not sure what would be better name for this, but using `authToken` might be confusing.

**Review by eyalbe4 — APPROVED** [review]

**Pull request merged**